### PR TITLE
Add loging

### PR DESCRIPTION
**Pull request by decobots**
#33

### CODE REVIEW

File: tests/decorators.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import logging
+from functools import wraps
+
+log = logging.getLogger()
+
+
+def log_test_name(func):
```
Comment (livercat) [inline tests/decorators.py]: Investigate pytest logging

File: tests/decorators.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+    @wraps(func)
+    def decorator(*args, **kwargs):
+        log.debug(f"==================================================== started {func.__name__}")
```
Comment (livercat) [inline tests/decorators.py]: ‘=‘ * 20 + f’text’

File: src/logger.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import logging
+import os
+
+
+def init_logging(filename):
+    log_path = os.path.join(os.path.dirname(os.path.dirname(__file__)), filename)
+    log = logging.getLogger()
+    handler = logging.FileHandler(filename=log_path)
+    formatter = logging.Formatter(
```
Comment (livercat) [inline src/logger.py]: Make format more readable

File: src/logger.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import logging
+import os
+
+
+def init_logging(filename):
+    log_path = os.path.join(os.path.dirname(os.path.dirname(__file__)), filename)
+    log = logging.getLogger()
+    handler = logging.FileHandler(filename=log_path)
```
Comment (livercat) [inline src/logger.py]: Consider adding a console handler

File: main.py [hunk 5]
```diff
@@ -1,21 +1,31 @@
...
+    init_logging("log.log")
     will_be_uploaded_N_photos = 3
 
     with DataBase(photos_table_name="photos_list") as db:
         twitter = Twitter(database=db)
         flickr = Flickr(database=db)
-        pictures = flickr.get_photos()
+
         for _ in range(will_be_uploaded_N_photos):
+            log.debug(f"==================================================== started cycle {_}")
```
Comment (livercat) [inline main.py]: Move these messages to methods

File: src/data_base.py [hunk 6]
```diff
@@ -33,33 +36,49 @@ def __exit__(self, *args, **kwargs):
...
+                sql.SQL("UPDATE {} SET posted = 'TRUE', tweet_id=%s WHERE id = %s").format(
+                    sql.Identifier(self.photos_table_name)), (post_id, photo_id))
+            log.info(f"photo with flickr id={post_id} updated, marked as posted to twitter")
 
     def unposted_photos(self) -> List[str]:
         self.cursor.execute(
             sql.SQL("SELECT id FROM {} WHERE posted = 'FALSE'").format(sql.Identifier(self.photos_table_name)))
-        return [row[0] for row in self.cursor.fetchall()]
+        result = [row[0] for row in self.cursor.fetchall()]
+        log.debug(f"in database {len(result)} unposted photos")
```
Comment (livercat) [inline src/data_base.py]: X unposted photos in database

File: src/flickr.py [hunk 7]
```diff
@@ -38,7 +44,7 @@ def get_photos(self) -> Dict[str, Photo]:
                           secret=tag.attrib["secret"],
                           title=tag.attrib["title"])
             result_photos[photo.id_flickr] = photo
-
+        log.info(f"received {len(result_photos)} from flickr")
```
Comment (livercat) [inline src/flickr.py]: “Photos” is missing after len()

File: src/request.py [hunk 8]
```diff
@@ -1,9 +1,13 @@
+import logging
 from xml.etree import ElementTree
 
 import requests
 
+log = logging.getLogger()
+
 
 def request(method_type: str, url: str, **kwargs):
+    log.debug(f"started request '{url}'")
```
Comment (livercat) [inline src/request.py]: “to url”

File: src/request.py [hunk 9]
```diff
@@ -16,11 +20,16 @@ def request(method_type: str, url: str, **kwargs):
                               data=data,
                               auth=kwargs.get("auth", None))
     if result.status_code != 200:
+        log.exception(f'request status code !=200 ({result.status_code})')
```
Comment (livercat) [inline src/request.py]: Response status code

File: src/request.py [hunk 10]
```diff
@@ -16,11 +20,16 @@ def request(method_type: str, url: str, **kwargs):
...
         raise ValueError("oops ", result.text)
+
     try:
         response_tag = ElementTree.fromstring(result.text)
+        log.debug(f'trying to parse as xml')
         if response_tag.attrib["stat"] == 'fail':
+            log.exception(f'server returned fail status in xml')
             raise ValueError(result.text)
+        log.debug(f'trying parsed as xml')
```
Comment (livercat) [inline src/request.py]: “Trying” is not needed: “Parsed xml”

File: src/request.py [hunk 11]
```diff
@@ -16,11 +20,16 @@ def request(method_type: str, url: str, **kwargs):
...
     except ElementTree.ParseError:
-        pass
+        log.debug(f'data received')
```
Comment (livercat) [inline src/request.py]: Is this a correct message?

File: src/flickr.py [hunk 12]
```diff
@@ -20,6 +23,9 @@ class Flickr:
     def __init__(self, database: DataBase, requester=request):
         self.request = requester
         self.db = database
+        logging.info(self.__class__)
```
[the remark this answers is not in the document]
Comment (decobots) [reply src/flickr.py]: Nope

File: src/logger.py [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+    formatter = logging.Formatter(
+        fmt="%(name)s ==>  %(asctime)s:%(levelname)s: %(module)s: %(funcName)s:  %(message)s")
+    log.addHandler(handler)
+    handler.setFormatter(formatter)
+    log.setLevel(logging.DEBUG)
+    logging.getLogger("urllib3").setLevel(logging.WARNING)
```
[the remark this answers is not in the document]
Comment (decobots) [reply src/logger.py]: done

File: tests/decorators.py (hunk 2 above)
Comment (decobots) [reply tests/decorators.py]: done

File: src/logger.py (hunk 3 above)
Comment (decobots) [reply src/logger.py]: done

File: main.py (hunk 5 above)
Comment (decobots) [reply main.py]: done

File: src/data_base.py (hunk 6 above)
Comment (decobots) [reply src/data_base.py]: done

File: src/flickr.py (hunk 7 above)
Comment (decobots) [reply src/flickr.py]: done

File: src/request.py (hunk 8 above)
Comment (decobots) [reply src/request.py]: done

File: src/request.py (hunk 9 above)
Comment (decobots) [reply src/request.py]: done

File: src/request.py (hunk 10 above)
Comment (decobots) [reply src/request.py]: done

File: src/request.py (hunk 11 above)
Comment (decobots) [reply src/request.py]: changed to 'request returned correct response'

File: tests/decorators.py (hunk 1 above)
Comment (decobots) [reply tests/decorators.py]: investigated, and I don't like it

File: src/request.py [hunk 14]
```diff
@@ -20,16 +20,19 @@ def request(method_type: str, url: str, **kwargs):
                               data=data,
                               auth=kwargs.get("auth", None))
     if result.status_code != 200:
-        log.exception(f'request status code !=200 ({result.status_code})')
+        log.exception(f'request status code !=200 (response status code {result.status_code})')
```
Comment (livercat) [inline src/request.py]: Response, not request

File: src/request.py [hunk 15]
```diff
@@ -20,16 +20,19 @@ def request(method_type: str, url: str, **kwargs):
...
         raise ValueError("oops ", result.text)
```
Comment (livercat) [inline src/request.py]: Oops?

File: src/request.py (hunk 14 above)
Comment (decobots) [reply src/request.py]: ok

File: src/request.py (hunk 15 above)
Comment (decobots) [reply src/request.py]: Do you know what means word 'fun' ?

**Pull request merged**